### DISCUSSION TITLE
How to compare string or date type?

### DISCUSSION THREAD

**Discussion by keispace** [fluree/db#719] [Q&A] [answered]
Hello, I have one more question.
While working on a web3 related project at the company, the concept and structure seem to be a very suitable DB for our project, so I have been testing it using docker image(fluree/server:latest) since last week

but, I can't find a way to compare strings, to be exact xsd:dataTime data type.

Here's my insert query:
```
{
  "@context":{
    "ex": "http://example.org/", 
    "xsd":"http://www.w3.org/2001/XMLSchema#", 
     "ex:date1":{"@type":"xsd:dateTime"} 
  },
  "ledger": "test/test",
  "insert":[
    {"@id":"24896510145","@type":"ex:CreateEvent","ex:str":"apple","ex:num":1,"ex:public":false,"ex:date1":"2024-01-30T15:00:00Z", "ex:date2":{"@type":"xsd:dateTime", "@value":"2024-01-30T15:00:00Z"}}
    ,{"@id":"24896510251","@type":"ex:PushEvent","ex:str":"cat","ex:num":2.3,"ex:public":true,"ex:date1":"2024-12-01T15:00:01Z", "ex:date2":{"@type":"xsd:dateTime", "@value":"2024-12-01T15:00:01Z"}}
]}
```

And I tried the query below ( ?dt1 and ?dt2 respectively)
But response is [].
```
{
  "@context":{
    "ex": "http://example.org/", 
    "xsd":"http://www.w3.org/2001/XMLSchema#", 
     "ex:date1":{"@type":"xsd:dateTime"} 
  },
  "from": "test/test",
  "where": [
    {
      "@id": "?s",
      "ex:date1":"?dt1",
      "ex:date2":"?dt2"
    }
    ,["filter","(> ?dt1 (now))"]
  ],
  "select":{"?s":["*"]}
}
```

Even if I change the filter to the content below, it will respond to [].(Normal when the filter is removed)
```
{
  "@context":{
    "ex": "http://example.org/", 
    "xsd":"http://www.w3.org/2001/XMLSchema#", 
     "ex:date1":{"@type":"xsd:dateTime"} 
  },
  "from": "test/test",
  "where": [
    {
      "@id": "?s",
      "ex:date1":"?dt1",
      "ex:date2":"?dt2"
    }
    ,["filter","(= ?dt1 \"2024-01-30T15:00:00Z\")"]
  ],
  "select":{"?s":["*"]}
}

```

Even if I look up the document, it's hard to find information about the manipulation of the date and string

So, I have two questions. 
1. Is there a way to compare string? ('a' <'b')
2. How to use a 'date type' in filter?
  - How to put data in to use `(< ?date(now))` of the document (followed the instructions but does not seem to work)
  - To filter data before and after a particular date/time/datetime

**Comment by thecacho** [comment]
Hi @keispace ! Good to see you again.

TL;DR: From what i know the Fluree database currently supports date/time predicates, but lacks direct date math functionalities. While there are functions to extract date components, using numerical comparisons for date operations is a bit clunky. The fluree team is looking into plans to improve this aspect in the future right now tho.

Could you please delve further into the first question? I couldn't understand it at all 🤔
> 1. Is there a way to compare strings? ('a' < 'b')
---

Resources
> Message taken from the Discord channel https://discord.com/channels/896089675511508992/1198700428427268107
> Wes (fluree core)— 22/01/2024 13:45
We do have support for date / time predicates (there are a few different ones defined in here): https://github.com/fluree/db/blob/347a28f8408eb74059f7144638e044cab00ec7b8/src/fluree/db/datatype.cljc#L15-L50
But no direct date math yet. I'll check on what the plans are for that.
The closest we get currently are the fns that extract the various date components (e.g. year, month, day) so you could construct this with numerical comparisons on those, but it would be kind of clunky. We'll hopefully improve on it down the road.
>

**Reply by keispace** [reply]
Hello @Domenic-MZS
Thank you for your help.

What I'm trying to do is look up data that has date data before/after a specific date.

Question 1, 

I tried to compare the dates in the dataset in the question, but it didn't work out, so I tried to compare the strings, but it didn't work either.

For example, in javascript or sql, '2024-02-05'>'2024-01-30' is `true`, so i tried bypass text comparison for the date.

```
{
 "@context":{
    "ex": "http://example.org/", 
    "xsd":"http://www.w3.org/2001/XMLSchema#", 
     "ex:date1":{"@type":"xsd:dateTime"} 
  },
  "from": "test/test",
  "where": [
    {
      "@id": "?s",
      "ex:date1":"?dt1",

    }
    ,["filter","(> ?dt1 \"2024-01-30T15:00:00Z\")"]

  ],
  "select":{"?s":["*"]}
}

```

The expected result was data including "ex:str":"cat" but the response is []


When I checked the db dependency of the docker image, I found that it specified a specific commit hash. I think I need to check if it's an improved version or not

**Reply by thecacho** [reply, marked as answer]
Oh i got it now 🙌🏼 💡.

The closest thing right now (as it was said) are the `date fns` (not mentioned on documentation yet).

With the `date-fns`  you will get a numeric value for each `year`, `month`, `day`, `hours`, `minutes` and `seconds`. To then write a `filter` for each one specifying that the `date` `year` should be `greater` than `x`, and/or the `month` should be greater and so on... 

> It's clunky. Because there's no direct math yet.  

You can see all the `date-fns` available over here:
https://github.com/fluree/db/blob/3dfce2907e5e2082a52b2ee8a5bc457f5aac9ac5/src/fluree/db/query/exec/eval.cljc#L295-L296

**Reply by thecacho** [reply]
However, this functionality seems unstable. I tried it with your example, and it failed with a `can't convert timezone to local timezone` error. But using a local timezone did work. Therefore, besides being the only path I see right now (apart from regex), this method is not just clunky but also bugged at the time. I would recommend you to use milliseconds since epoch to store dates and filter it more easily (handling only numbers) or wait until direct date-math is implemented.
